src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/LocalDBAPI.Windows.cs

```diff
@@ -3,300 +3,167 @@
-        }
-
-
-        static SqlException CreateLocalDBException(string errorMessage, string instance = null, int localDbError = 0, uint sniError = 0)
-        {
-            Debug.Assert((localDbError == 0) || (sniError == 0), "LocalDB error and SNI error cannot be specified simultaneously");
-            Debug.Assert(!string.IsNullOrEmpty(errorMessage), "Error message should not be null or empty");
-            SqlErrorCollection collection = new SqlErrorCollection();
-
-            int errorCode = (localDbError == 0) ? (int)sniError : localDbError;
-
-            if (sniError != 0)
-            {
-                string sniErrorMessage = SQL.GetSNIErrorMessage(sniError);
-                errorMessage = String.Format((IFormatProvider)null, "{0} (error: {1} - {2})",
-                         errorMessage, sniError, sniErrorMessage);
-            }
-
-            collection.Add(new SqlError(errorCode, 0, TdsEnums.FATAL_ERROR_CLASS, instance, errorMessage, null, 0));
-
-            if (localDbError != 0)
-                collection.Add(new SqlError(errorCode, 0, TdsEnums.FATAL_ERROR_CLASS, instance, GetLocalDBMessage(localDbError), null, 0));
-
-            SqlException exc = SqlException.CreateException(collection, null);
-
-            exc._doNotReconnect = true;
-
-            return exc;
-        }
-
-        private class InstanceInfo
-        {
-            internal InstanceInfo(string version)
-            {
-                this.version = version;
-                this.created = false;
-            }
-
-            internal readonly string version;
-            internal bool created;
-        }
-
-        static object s_configLock = new object();
-        static Dictionary<string, InstanceInfo> s_configurableInstances = null;
-
-        internal static void DemandLocalDBPermissions()
+        
+        private static IntPtr UserInstanceDLLHandle
         {
-            if (!_partialTrustAllowed)
+            get
             {
-                if (!_partialTrustFlagChecked)
+                if (s_userInstanceDLLHandle == IntPtr.Zero)
                 {
-                    object partialTrustFlagValue = AppDomain.CurrentDomain.GetData(Const_partialTrustFlagKey);
-                    if (partialTrustFlagValue != null && partialTrustFlagValue is bool)
-                    {
-                        _partialTrustAllowed = (bool)partialTrustFlagValue;
-                    }
-                    _partialTrustFlagChecked = true;
-                    if (_partialTrustAllowed)
+                    #if NETFRAMEWORK
+                    RuntimeHelpers.PrepareConstrainedRegions();
+                    #endif
+                    
+                    lock (s_dllLock)
                     {
-                        return;
+                        if (s_userInstanceDLLHandle == IntPtr.Zero)
+                        {
+                            SniNativeWrapper.SNIQueryInfo(QueryType.SNI_QUERY_LOCALDB_HMODULE, ref s_userInstanceDLLHandle);
```

Comment:
Here's a comparison of the versions of that specific property
![image](https://github.com/user-attachments/assets/f8290e5a-f08f-49bc-bbcf-457e571c5591)

As far as I can tell, the only real difference between netfx and netcore is that netfx is manually using a Monitor for the lock whereas netcore is using a `lock` block. Also netcore uses Kernel32.GetProcAddress whereas netfx uses Kernel32Safe.GetProcAddress in LoadProcAddress.

My one question is, since there isn't a try/catch/finally anymore, is it still required to use the `PrepareConstrainedRegion`?

You did however point out one bug that I will fix shortly - both `LocalDBCreateInstance` and `LocalDbFormatMessage` were calling `LoadProcAddress`. It uses `Kernel32[Safe].GetProcAddress` with `UserInstanceDLLHandle` and a hardcoded function name. I'll change it to get the func name as a parameter. 

As for handle leaks ... can you explain more? I'm just pulling the code over from one project to another, without any functionality changes. If it *is* a leak, we could open that as a separate issue.

Reply:
It looks like netcore tries to load the same DLL via two different methods. That's... a choice...

You're correct, the `UserInstanceDLLHandle` is a fair merge. In netcore, this is only used to get the delegate for `LocalDBAPI.LocalDBFormatMessage` though. `LocalDB.GetLocalDBConnectionString` starts the LocalDB instance by calling `LocalDB.LoadUserInstanceDll`, and this is the method which performs the `LoadLibraryExW`/`GetProcAddress` dance which I'm talking about.

In netfx, the closest equivalent is the `LocalDBAPI.LocalDBCreateInstance` property, which uses the `UserInstanceDLLHandle` property.

wrt the handle leak: `s_userInstanceDLLHandle` is an IntPtr which (I believe) comes from a LoadLibrary call. If it was a SafeLibraryHandle, `FreeLibrary` would be called when the handle was released or finalised. The handle is kept alive because it's a field on the static `LocalDB.Instance`. In principle, if SqlClient is unloaded (via an AssemblyLoadContext) I expect that static instance to be finalised, the SafeLibraryHandle to be finalised and FreeLibrary to be called. If it's an IntPtr, when SqlClient is unloaded and LocalDB.Instance is finalised, the IntPtr gets discarded and the handle it contains is left dangling.